# ML Assignment 1
Iciar Adeliño - March 2026 - Machine Learning Foundations

## Task Ordering

1. **Identifying the prediction target**:
   We must first define what we are predicting. The target determines the task type, evaluation metrics, and which features are valid at prediction time.

2. **Data loading and exploration**:
   This step is descriptive and safe before splitting. It helps us understand variable types, distributions, missing values, and potential class imbalance without fitting any transformations.

3. **Data splitting (stratified)**:
   Splitting must occur before any transformation that learns from data to prevent leakage. Train, validation, and test sets must remain statistically independent.

4. **Managing missing values**:
   Imputation relies on distributional statistics, so it must be fitted on the training set and then applied to validation and test sets.

5. **Encoding categorical variables**:
   Future steps require numerical input.

6. **Feature scaling**:
   Scaling improves optimization and coefficient comparability.

7. **Feature selection**:
   Now that the featurs are finalized, we can select the relvant ones for our model. 

8. **Addressing class imbalance**:
   Resampling must operate on the fully engineered feature space, so we do it after all feature engineering.

9. **Training**:
   We train the model on the processed training set, validate performance on the validation set, and use the test set only once for final evaluation.


In [5]:
# import everything at the top

import pandas as pd

## Identifying prediction target 

In [6]:
df = pd.read_csv("bank-additional.csv", sep=";")
print(df.head())

   age          job  marital          education default  housing     loan  \
0   30  blue-collar  married           basic.9y      no      yes       no   
1   39     services   single        high.school      no       no       no   
2   25     services  married        high.school      no      yes       no   
3   38     services  married           basic.9y      no  unknown  unknown   
4   47       admin.  married  university.degree      no      yes       no   

     contact month day_of_week  ...  campaign  pdays  previous     poutcome  \
0   cellular   may         fri  ...         2    999         0  nonexistent   
1  telephone   may         fri  ...         4    999         0  nonexistent   
2  telephone   jun         wed  ...         1    999         0  nonexistent   
3  telephone   jun         fri  ...         3    999         0  nonexistent   
4   cellular   nov         mon  ...         1    999         0  nonexistent   

  emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  nr.e

It is useful to understand what each column measures (found with some research):

| variable name    | role    | type        | description |
|------------------|---------|-------------|-------------|
| `age`            | feature | integer     | age of the client |
| `job`            | feature | categorical | type of job (admin., blue-collar, entrepreneur, housemaid, management, retired, self-employed, services, student, technician, unemployed, unknown) |
| `marital`        | feature | categorical | marital status: married, divorced, single, unknown |
| `education`      | feature | categorical | education level: basic.4y, basic.6y, basic.9y, high.school, illiterate, professional.course, university.degree, unknown |
| `default`        | feature | binary      | has credit in default? (yes / no / unknown) |
| `housing`        | feature | binary      | has a housing loan? (yes / no / unknown) |
| `loan`           | feature | binary      | has a personal loan? (yes / no / unknown) |
| `contact`        | feature | categorical | contact communication type: cellular or telephone |
| `month`          | feature | categorical | last contact month of the year (jan – dec) |
| `day_of_week`    | feature | categorical | last contact day of the week (mon – fri) |
| `duration`       | feature | integer     | last contact duration in seconds, only known after the call ends |
| `campaign`       | feature | integer     | number of contacts made to this client during this campaign |
| `pdays`          | feature | integer     | days since client was last contacted from a previous campaign (999 = never contacted before) |
| `previous`       | feature | integer     | number of contacts made to this client before this campaign |
| `poutcome`       | feature | categorical | outcome of the previous campaign: nonexistent, failure, success |
| `emp.var.rate`   | feature | continuous  | employment variation rate, quarterly macro indicator |
| `cons.price.idx` | feature | continuous  | consumer price index, monthly macro indicator |
| `cons.conf.idx`  | feature | continuous  | consumer confidence index, monthly macro indicator |
| `euribor3m`      | feature | continuous  | euribor 3-month interest rate, daily macro indicator |
| `nr.employed`    | feature | continuous  | number of employees in the economy — quarterly macro indicator |
| `y`              | target  | binary      | has the client subscribed to a term deposit? (yes / no) |

The target variable is `y`. It indicates whether a client subscribed to a term deposit (`"yes"` or `"no"`). It is the appropriate prediction objective for this marketing campaign for the following reasons:

- The entire campaign exists to identify clients likely to subscribe to a term deposit, which what `y` directly encodes.
- `y` is not observed until the client makes their decision at the end of the call, it is the event we are trying to anticipate.
- There are only two values for `y` which makes sense, knowing that what we are trying to predict is whether they subscribe or not. 

Variables that could superficially appear to be valid targets:

1. `duration` — last contact duration in seconds

Call duration is strongly correlated with the subscription outcome, as longer calls tend to result in a subscription. It seems to reflect the outcome of the interaction. However, it is not exactly what we are trying to precict. It is not binary and a longer call does not necessarily mean a successful subscription.

2. `poutcome` — outcome of the previous marketing campaign

This variable encodes success/failure so it might look like what we have to predict. However, `poutcome` refers to a past campaign, not the current one, which is not our objective.

## Data loading and exploration 

## Data splitting

##